### PR TITLE
added support for compiling source with -flto option disabled.

### PR DESCRIPTION
**Pull request by mtei**
## Description

Some source code (eg helix-serial.c) is sensitive to execution timing and will not work as expected when compiled with the -flto option.

This commit adds the ability to specify the -fno-lto option for specific source code in tmk_core/rules.mk.

### Usage:

file: keybaord, keymap, others rules.mk
```
SRC += timing_sensitive_code.c
```

change to
```
SRC += timing_sensitive_code.c/NO-LTO
```

~~This commit does not change the build results.~~

### Postscript
Currently there are keyboards that use LIB_SRC or QUANTUM_LIB_SRC to specify -fno-lto. But this is not correct.

* LIB_SRC and QUANTUM_LIB_SRC are ways to specify that objects are stored in their respective libraries.
* LIB_SRC and QUANTUM_LIB_SRC **ARE NOT** a way to specify -fno-lto for compilation.

~~Currently, the file specified in LIB_SRC and QUANTUM_LIB_SRC is compiled with -fno-lto, but this is only a workaround for the 'ar' command not supporting lto object. If the 'ar' command supports lto objects, this action is gone.~~

With this PR change, LIB_SRC, QUANTUM_LIB_SRC no longer means adding `-fno-lto`.

## Types of Changes

- [x] Core
- [ ] Bugfix
- [ ] New feature
- [x] Enhancement/optimization
- [ ] Keyboard (addition or update)
- [x] Keymap/layout/userspace (addition or update)
- [ ] Documentation

## Checklist

- [x] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [x] I have read the [**CONTRIBUTING** document](https://docs.qmk.fm/#/contributing).
- [ ] I have added tests to cover my changes.
- [x] I have tested the changes and verified that they work and don't break anything (as well as I can manage).

### CODE REVIEW

**Comment by zvecr** [comment]
@mtei as we currently use `QUANTUM_LIB_SRC` to avoid the issues you have highlighted, can you give some insight to the value add of this PR. It would be good to have that context for reviewing the changes.

**Comment by mtei** [comment]
Of course, I know QUANTUM_LIB_SRC and LIB_SRC. Because it was my contribution. (#4522)

LIB_SRC is for automatically selecting whether a link is required. Certainly, the source specified by LIB_SRC is compiled with -fno-lto. But that's just a side effect. It is not the main purpose of LIB_SRC.

I believe there must be a way to explicitly compile with -fno-lto.

For example, when `LINK_TIME_OPTIMIZATION_ENABLE=no`, the following two statements have exactly the same result. So if you are sure that it already works, you can safely add `/NO-LTO`.

```
SRC += a.c b.c c.c
```
```
SRC += a.c b.c/NO-LTO c.c
```

**Comment by drashna** [comment]
>LIB_SRC is for automatically selecting whether a link is required. Certainly, the source specified by LIB_SRC is compiled with -fno-lto. But that's just a side effect. It is not the main purpose of LIB_SRC.
>
>I believe there must be a way to explicitly specify compiling with -fno-lto.

Is there a specific/technical reason that `LIB_SRC` won't work for this?  

I mean, if `LIB_SRC` is doing the same thing, and some extra stuff that doesn't cause issues, I don't see why we can't use that over adding another setting here.    Because adding more options like this feels like it may be adding complexity for edge cases where other options would not only work, but work well.  Which will add confusion down the line.

**Comment by mtei** [comment]
Currently, the source specified by LIB_SRC is compiled with `-fno-lto` because  'tmk_core/rules.mk#L37' forces it.

https://github.com/qmk/qmk_firmware/blob/f64d9b06215bb08d7f77aeba126c0804fffd0064/tmk_core/rules.mk#L37

If we comment out this, it will look like this:

```
{qmk_firmware} % make LINK_TIME_OPTIMIZATION_ENABLE=yes  lets_split:default
QMK Firmware 0.7.40
Making lets_split/rev2 with keymap default

avr-gcc (GCC) 8.3.0
Copyright (C) 2018 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

Compiling: keyboards/lets_split/lets_split.c                                                        [OK]
......
Compiling: quantum/split_common/serial.c                                                            [OK]
Archiving: .build/obj_lets_split_rev2_default/quantum/split_common/serial.o                         [WARNINGS]
 | 
 | avr-ar: .build/obj_lets_split_rev2_default/quantum/split_common/serial.o: plugin needed to handle lto object
 | 
Compiling: drivers/avr/i2c_master.c                                                                 [OK]
Archiving: .build/obj_lets_split_rev2_default/i2c_master.o                                          [WARNINGS]
 | 
 | avr-ar: .build/obj_lets_split_rev2_default/i2c_master.o: plugin needed to handle lto object
 | 
Compiling: drivers/avr/i2c_slave.c                                                                  [OK]
Archiving: .build/obj_lets_split_rev2_default/i2c_slave.o                                           [WARNINGS]
 | 
 | avr-ar: .build/obj_lets_split_rev2_default/i2c_slave.o: plugin needed to handle lto object
 | 
Compiling: tmk_core/common/host.c                                                                   [OK]
......
Linking: .build/lets_split_rev2_default.elf                                                         [ERRORS]
```

As you can see in the message above, avr-ar does not currently support lto objects.

In the future, if avr-ar will support lto objects as standard, it will not be necessary to add `-fno-lto` in 'tmk_core/rules.mk'.
In other words, using LIB_SRC does not guarantee to use `-fno-lto`.

This is why we need a way to explicitly specify `-fno-lto`.

**Comment by drashna** [comment]
To be blunt, I don't feel that your answer answers my question, but skirts around it. 

>If we comment out this, it will look like this:

Why would we comment that out?   Aside from as an example.

Though a better question may be, what are `LIB_SRC` and `QUANTUM_LIB_SRC` doing that different that they shouldn't be used for disabling LTO? 


Basically, my concern here is about adding more and more options to the makefile stuff, and added complexity here, without clear definition of which each option does, in detail.  

In fact, these different `SRC` variables should probably be documented proper in:
https://github.com/qmk/qmk_firmware/blob/master/docs/config_options.md#build-options

**Comment by mtei** [comment]
> > If we comment out this, it will look like this:
> 
> Why would we comment that out? Aside from as an example.

To show the problem I first encountered when I implemented LIB_SRC.

To avoid this problem, I had to use the -fno-lto option. This is only a workaround. The original role of LIB_SRC is to store objects in their respective libraries. 
(I added explanation to 'Description' today, so please read it again.)

So if the problem is gone, we don't need to use the -fno-lto option in our LIB_SRC implementation.

**Comment by mtei** [comment]
I added a description of LIB_SRC to config_options.md.

**Comment by drashna** [comment]
Okay, to see if I understand this correctly...

The `LIB_SRC` and `QUANTUM_LIB_SRC`, add the files after `SRC` and `QUANTUM_SRC` (I think).  Additionally, if you have the same file added multiple times to the `LIB` version, it parses that correctly, so it doesn't generate a warning about multiple inclusions.    

This makes the compilation cleaner, and more intelligent.  

RIght now, those also assume `no lto`, but that's more of an unintentional side effect, rather than intended functionality. 

Ideally, stuff like the `i2c_master.c` should  be called by using `LIB_SRC += i2c_master.c/NO-LTO`.

If that's correct, awesome, and I think this would be a good change.   I was just having a hard time understanding why this was a good change, or why it was needed. 

Thank you for taking the time and effort to explain this!

**Comment by mtei** [comment]
I'm glad you understood.

> Okay, to see if I understand this correctly...
> 
> The `LIB_SRC` and `QUANTUM_LIB_SRC`, add the files after `SRC` and `QUANTUM_SRC` (I think). Additionally, if you have the same file added multiple times to the `LIB` version, it parses that correctly, so it doesn't generate a warning about multiple inclusions.

That is correct.

In addition, in the following example, lib_d.a is linked if any of a.o, b.o or c.o calls a function in lib_d.a. Otherwise, lib_d.a will not be linked.

```
# file: rules.mk
LIB_SRC += lib_d.c
SRC += a.c
SRC += b.c
SRC += c.c
```

The link order is as follows.
```
...  a.o b.o c.o  ...  lib_d.a  ...
```

**Comment by mtei** [comment]
> Ideally, stuff like the `i2c_master.c` should be called by using `LIB_SRC += i2c_master.c/NO-LTO`.

`/NO-LTO` can be specified for both SRC and LIB_SRC files, as shown in the following example.

* common_features.mk#L387-L389
  ```c
        QUANTUM_LIB_SRC += $(QUANTUM_DIR)/split_common/serial.c/NO-LTO \
                           i2c_master.c \
                           i2c_slave.c
  ```
* keyboards/helix/rev2/rules.mk#L3-L6
  ```c
  SRC += local_drivers/i2c.c
  SRC += local_drivers/serial.c/NO-LTO
  SRC += local_drivers/ssd1306.c
  KEYBOARD_PATHS += $(HELIX_TOP_DIR)/local_drivers
  ```

**Comment by mtei** [comment]
Changed to use `gcc-ar` instead of `ar` command for library creation. Objects compiled with LTO enabled can now be stored in a library. As a result, the workaround in `tmk_core/rules.mk` is no longer needed and has been removed.

Currently, `LIB_SRC` and `QUANTUM_LIB_SRC` no longer mean adding `-fno-lto`.

https://gcc.gnu.org/onlinedocs/gcc/Optimize-Options.html#index-ffat-lto-objects

**Comment by noroadsleft** [comment]
This has merge conflicts.

**Comment by drashna** [comment]
Now that I've looked at this closer
>With this PR change, LIB_SRC, QUANTUM_LIB_SRC no longer means adding -fno-lto.

If this is the case, then the places where `QUANTUM_LIB_SRC += i2c_master.c` is called should be moved.

But it sounds like we don't have a `LIB` option, for when multiple things include the same file (`i2c_master.c` namely).  If that's the case, it will generate warnings for anything that may trigger this (such as using OLEDs and rgb matrix with an ISSI controller, or the ergodox ez glow). 

If that's the case, that may be a blocker to getting this merged

**Comment by mtei** [comment]
I'd like to break this PR down into a few PRs. This should make it easier to verify each one.

I'm going to do it when I have more time.

**Comment by Erovia** [comment]
@mtei, would you like to leave this PR open or can it be closed?

**Comment by mtei** [comment]
> @mtei, would you like to leave this PR open or can it be closed?

Now I'm getting ready to split up this PR and reopen a new one.

Hold on a moment.

**Comment by tzarc** [comment]
Converted to draft so that someone doesn't decide to have a bright idea and attempt to merge.
